src/ngraph/runtime/cpu/pass/cpu_rnn_mat_fusion.cpp

```diff
@@ -0,0 +1,288 @@
+                FindValidSegments(
+                    out_node, segment, segment_bundle, valid_sequence_list, depth + 1);
+            }
+        }
+    }
+}
+
+struct OrderedParams
+{
+public:
+    OrderedParams()
+        : m_params{{nullptr, nullptr, nullptr}}
+    {
+    }
+
+    bool exist(const NodeSegment::Type type) { return m_params[type] != nullptr; }
+    bool valid()
+    {
+        return std::none_of(m_params.cbegin(), m_params.cend(), [](const NodePtr& n) -> bool {
+            return n == nullptr;
+        });
+    }
+
+    void set(const NodeSegment::Type type, const NodePtr& node) { m_params[type] = node; }
+    NodePtr get(const NodeSegment::Type type) const { return m_params.at(type); }
+    friend bool operator<(const OrderedParams& a, const OrderedParams& b);
+
+private:
+    // order based on NodeSegment::Type
+    // <data, weights, bias>
+    std::array<NodePtr, 3> m_params;
+};
+
+bool operator<(const OrderedParams& a, const OrderedParams& b)
+{
+    return a.m_params < b.m_params;
+}
+
+bool runtime::cpu::pass::CPURnnMatFusion::run_on_function(std::shared_ptr<Function> function)
+{
+    bool modified = false;
+
+    const NodeTypeSequenceList valid_sequences{
+        {NodeSegment::DATA,
+         {TI(op::Parameter), TI(op::Slice), TI(op::Reshape), TI(op::Dot), TI(op::Add)}},
+        {NodeSegment::WEIGHTS, {TI(op::Parameter), TI(op::Reshape), TI(op::Dot), TI(op::Add)}},
+        {NodeSegment::BIAS, {TI(op::Parameter), TI(op::Broadcast), TI(op::Add)}}};
+    // this is the expected sequence count for fusing
+    const size_t segment_count = 3;
+
+    // find all parameter nodes
+    std::vector<NodePtr> param_nodes;
+    for (auto& node : function->get_ordered_ops())
+    {
+        const Node& node_ref = *node;
+        if (TI(node_ref) == TI(op::Parameter))
+        {
+            param_nodes.push_back(node);
+        }
+    }
+
+    // iterate all parameters and find all valid segments
+    std::vector<NodeSegment> segment_bundle;
+    for (auto& node : param_nodes)
+    {
+        NodeSegment segment;
+        FindValidSegments(node, segment, segment_bundle, valid_sequences, 0);
+    }
+
+    // combined all segments by last operator
+    std::map<NodePtr, std::vector<NodeSegment>> op_seg_map;
+    for (const auto& segment : segment_bundle)
+    {
+        auto op_it = op_seg_map.find(segment.back());
+        if (op_it == op_seg_map.end())
+        {
+            auto insert_result = op_seg_map.insert(
+                std::make_pair(segment.back(), std::vector<NodeSegment>(segment_count)));
+            op_it = insert_result.first;
+        }
+        (op_it->second)[segment.type] = segment;
+    }
+
+    // remove ops with less than segment_count number of segments
+    for (auto op_it = op_seg_map.cbegin(); op_it != op_seg_map.cend();)
+    {
+        // remove ops with less than expected segements
+        bool valid = true;
+        for (auto& seg : op_it->second)
+        {
+            if (seg.empty())
```

Comment:
I think this needs a comment. This seems to mean a segment of a particular type is missing (kinda `nullptr`) so that slot is initialized w/ a default segment.